### DISCUSSION TITLE
Does Entity need to know all possible components?

### DISCUSSION THREAD

**Discussion by JKarneeb** [hmans/miniplex#295] [Q&A] [answered]
Based on the examples, it looks like the base Entity type that all entities of a World have expects it to enumerate all of the possible components it could have? Is this right?

I suppose in practice, its not that big of a deal to enumerate all of the possible types, but it feels odd from an ECS perspective. I like how lightweight this library is compared to libraries that are more hyper-focused on the performance aspect of ECS, but I can't shake the feeling that having a monolithic object that can store 1 of each possible component just feels off.

How do you all tend to use your Entity class? Do you leave it totally untyped? Fully specified? What is the expected best practice here?

**Comment by hamzakubba** [comment]
The library is fully typed (it infers correctly in the relevant use cases), I'm not the author but I believe that was the hardest part of building the library. Did you open and try to edit the example code in a typescript-capable editor?

**Reply by JKarneeb** [reply]
Yes I understand that it's typed. I meant more so that the way that it seems to want to handle typing for entities specifically is to specify an interface for Entity that has an optional field for every possible component in your system. Effectively the Entity class enumerates all components in it's specification.

I can't think of a particularly elegant way to handle it otherwise, I was just wondering if most people who use the library tend to fully spec their entity class, as once you have a large number of components that Entity class is going to be quite verbose. Unless I'm missing something?

**Reply by hamzakubba** [reply]
I'm not sure I understand the concern; you can do a lot to organize things if the concern is about organization, e.g.
```ts
// one way, break into sub-types or interfaces and compose them
type Instanced = {
  instanced: {
    // ...
  }
}
type Entity = Partial<Instanced & AnotherThing & YetAnotherThing>

// another way:
interface Entity {
  instanced?: InstancedComponent;
  anotherThing?: AnotherComponent;
  yetAnotherThing?: YetAnotherComponent;
}
```
You don't need an Entity class, it's just types. If specifying types seems like a downside overall for your project then that project can just be vanilla JS. I have been putting most types in a single file which has zero implementation, and that has worked well for me. In my case most things are inline, whereas a few are pulled out into their own interfaces (like the second example above).

**Reply by hmans** [reply]
It feels to me that the original question is not as much a question about how Miniplex expects entity types to be set up as much as it is about the overall design of the library (@JKarneeb, please correct me if I'm wrong.)

Miniplex chooses to express entities as plain JavaScript objects, with components being properties on those objects. This has the obvious limitation that you need to know in advance which components you want them to support if you want to prevent namespace clashes (you can't have two different components of the same name, because they'd fight for the same property name within entity objects.) This is why the up-front typing might feel a little awkward, but it's really just a function of you having to make decisions about what components to support.

Other (more "true") ECS implementations do not have this issue because there typically isn't an entity object, but just a collection of independent arrays (or other structures) storing component data, with the entity merely acting as the ID within those arrays.

Miniplex chooses this design because its primary intention is to be _simple to use_, especially for users who're familiar with state store patterns in JS, but new to ECS thinking. It's an extremely short jump from something like Zustand or Statery to Miniplex, and for many people who enjoy using Miniplex, that's one of the major reasons why.

**Reply by JKarneeb** [reply]
You are mostly correct. My intention wasn't to complain or to claim there was an issue, but more so to feel out how larger (non-demo projects) tend to handle this particular aspect.  It isn't particularly unreasonable for there to be upwards of 50+ component types in a game, depending on how you use them and how large your game is, especially if you make heavy use of tags. On top of that simply being a rather verbose object, it also means that wherever you define your Entity and World objects, you need to import every component type definition you plan to use. I'm a strong believer in decoupled code logic -- my intuition is to separate the initialization of things like my engine from specific mechanics. Front-loading everything its just counter to my general design style.

I think you mostly answered my question though @hmans. Miniplex is designed where it is expected that you will up-front type your Entity/World for the benefit of ease of use throughout the rest of the code-base.  I'm currently still just prototyping various patterns in the system to decide if miniplex is right for my coding style -- however there is no denying that it is extremely simple to use and provides many of the design benefits that ECS grants while still allowing the developer extreme freedom. This is something I value a lot, as I've found that, especially in THREE.js, trying to do pure ECS is a bit nightmarish. At least so far, I've not run into those issues with miniplex (but I'm still early in the honeymoon phase of a new tool heh)

Additionally, yes, the nature of the way Entity is structured implies that you cannot by default have duplicates of components, but I think that can be designed around easily enough.

**Reply by hmans** [reply, marked as answer]
So, one thing to remember is that there are many ways to actually compose the Entity type, instead of defining it monolithically. I've had projects where I did something like the following:

```ts
type IPosition {
  position: Vector3
}

type IHealth {
  health: { current: number, max: number }
}

type Entity = Partial<IPosition & IHealth & ...>
```

There are, of course, many variations on this.

In a similar vein, remember that Miniplex expects entities to be objects, but doesn't really care what these objects actually are. I've had at least one project where I actually implemented entities as _class instances_ to explore this. It might feel far removed from typical ECS thinking, but I was enjoying it as yet another nice pattern bridging the gap between the relaxed, classy JS world, and ECS design.

There's probably lots of fun and/or interesting patterns to discover!

**Comment by duchu-net** [comment]
Old topic, but recently I've been prototyping simulation engine inspired by nestjs, I was looking for an answer to a similar issue, because my Entity was always swelling, dependency growing... Coming to similar conclusions, now each entity is constructed at point of use, so I have good insight into dependencies between systems and components.

Idea was modularity, so that systems could be run in different configurations, overrides, or combinations, maybe that variation will be useful to someone ;)

```
- core
  - engine.ts
- common
  - decorators
    - injectable.decorator.ts
    - event-handler.decorator.ts
  - world
    - world.service.ts // extended miniplex World, Services are essentially Systems, just without `execute` method - more like data/logic providers
- modules
  - movement
    - target.component.ts
    - movement.system.ts
  - spatial
    - parent.component.ts
    - transform.component.ts
    - spatial.system.ts
```
```ts
// eg. movement.system.ts ----------
type MovementEntity = {
  id: string;
  [TransformComponent.componentKey]?: TransformComponent;
  [TargetComponent.componentKey]?: TargetComponent;
  [ParentComponent.componentKey]?: ParentComponent;
};
type WithTransform = With<MovementEntity, "transform">;

@Injectable() // registers system in Engine injectable api
export class MovementSystem implements XSystem {
  private query: Query<With<MovementEntity, "transform" | "target">>;
  private speed = 50000; // 50km/s

  constructor(
    private world: WorldService<MovementEntity>, // autoresolved by Engine
    private spatialSystem: SpatialSystem
  ) {
    this.query = world.with("transform", "target");
  }

  execute(deltaTime: number): void {
    for (const { target, transform, parent } of this.query) {
      const targetEntt = this.world.byId<WithTransform>(target.targetId);
      // ...
      transform.setMatrix(...)
      // ...
    }
  }

  @EventHandler("test_event_123") // registers method as event handler in Engine
  someTestHandler(event: any) {
    // ...
  }
}

// eg. transform.component.ts ----------
export type SerializedTransformComponent = {
  matrix: Matrix4Tuple;
};

export class TransformComponent implements XComponentSync {
  static componentKey = "transform" as const; // static name to be used as key in entity object
  updated = true; // when true, other system will send update to user and set it to false

  constructor(
    public matrix: Matrix4Tuple = new Matrix4().identity().toArray()
  ) {}

  setMatrix(matrix: Matrix4) {
    this.matrix = matrix.toArray();
    this.updated = true;
    return this;
  }

  deserialize(data: SerializedTransformComponent): void {}
  serialize(): SerializedTransformComponent {}
}
```

**Reply by JKarneeb** [reply]
Hmm, am I understanding this correctly that you are using multiple worlds? Doesn't world still need to be created with the full entity spec?

I see you are taking in a world service that is already types to your corresponding entity type, are you using that as a partial here and your main world is fully specified?

I'd need to go look at my old project again, but I ended up defining a bunch of individual objects and then at the very top scope of my project fully instantiating the entire object as a combination of all those partials. Curious how your approach differs
